backuptar/tar.go

```diff
@@ -317,32 +319,34 @@ func FileInfoFromHeader(hdr *tar.Header) (name string, size int64, fileInfo *win
 // tar file that was not processed, or io.EOF is there are no more.
 func WriteBackupStreamFromTarFile(w io.Writer, t *tar.Reader, hdr *tar.Header) (*tar.Header, error) {
 	bw := winio.NewBackupStreamWriter(w)
-	var sd []byte
+	var sd *windows.SECURITY_DESCRIPTOR
 	var err error
 	// Maintaining old SDDL-based behavior for backward compatibility.  All new tar headers written
 	// by this library will have raw binary for the security descriptor.
 	if sddl, ok := hdr.Winheaders[hdrSecurityDescriptor]; ok {
-		sd, err = winio.SddlToSecurityDescriptor(sddl)
+		sd, err = windows.SecurityDescriptorFromString(sddl)
 		if err != nil {
 			return nil, err
 		}
 	}
 	if sdraw, ok := hdr.Winheaders[hdrRawSecurityDescriptor]; ok {
-		sd, err = base64.StdEncoding.DecodeString(sdraw)
+		sdbytes, err := base64.StdEncoding.DecodeString(sdraw)
+		sd = (*windows.SECURITY_DESCRIPTOR)(unsafe.Pointer(&sdbytes[0]))
 		if err != nil {
 			return nil, err
 		}
 	}
-	if len(sd) != 0 {
+	sdLen := sd.Length()
```

Comment:
This seems risky to me. I think we are effectively trusting the security descriptor (from a tar archive) to tell us its own length, and then doing a copy equal to that length. This could result in malicious input causing us to leak data to the file, or crash the process.

Reply:
I think for the `hdrRawSecurityDescriptor` we should stick to the old code that treats it as a `[]byte`. Then we have a reliable length.

For the `hdrSecurityDescriptor` case, we might need to keep the old function around, as `windows.SecurityDescriptorFromString` doesn't return the length of the data it wrote out.